### PR TITLE
Add support for PETSc solvers

### PR DESCRIPTION
**Pull request by guyer**
Addresses #387

TODO:
  - [x] test in parallel on Py3k
  - [x] revise test matrix
  - [x] document scaling
    This situation is unclear and needs more analysis.
  - [x] verify threads vs. cpus
    This situation is unclear and needs more analysis.
  - [x] disable electrochem tests
  - [x] sort out precision issues
  - [x] `nproc` > 2
     Failures in
       - ~~`Doctest: fipy.meshes.gmshMesh.Gmsh2D`~~
       - ~~`Doctest: examples.diffusion.nthOrder.input4thOrder1D`~~
       - ~~`Doctest: fipy.meshes.representations.gridRepresentation._GridRepresentation._test`~~
       - `Doctest: examples.levelSet.distanceFunction.square` (due to #684, which is beyond the scope of this PR)
  - [x] revise "Running Under Python 3"      
  - [x] update fipy-feedstock
  - [x] update `environment.yml`/`requirements.txt` (where are these?!? Ah, in `binder/`.)

### CODE REVIEW

File: fipy/solvers/petsc/linearLUSolver.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        """
+        PETScSolver.__init__(self, tolerance=tolerance,
+                             iterations=iterations, precon="lu")
+
+    def _solve_(self, L, x, b):
+        ksp = PETSc.KSP()
+        ksp.create(PETSc.COMM_WORLD)
+        ksp.setType("preonly")
+        ksp.getPC().setType(self.preconditioner)
+        # TODO: SuperLU invoked with PCFactorSetMatSolverPackage(pc, MATSOLVERSUPERLU)
```
Comment (tkphd) [inline fipy/solvers/petsc/linearLUSolver.py]: Is this a todo-TODO, or an in the fullness of time TODO?

File: fipy/solvers/petsc/petscKrylovSolver.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+           -8: "KSP_DIVERGED_INDEFINITE_PC",
+           -9: "KSP_DIVERGED_NANORINF",
+           -10: "KSP_DIVERGED_INDEFINITE_MAT",
+           -11: "KSP_DIVERGED_PC_FAILED",
+           0: "KSP_CONVERGED_ITERATING"}
+
+class PETScKrylovSolver(PETScSolver):
+
+    """
+    .. attention:: This class is abstract, always create on of its subclasses.
```
Comment (tkphd) [inline fipy/solvers/petsc/petscKrylovSolver.py]: create one of

File: fipy/matrices/petscMatrix.py [hunk 3]
```diff
@@ -0,0 +1,805 @@
...
+            ids = numerix.arange(self._shape[0])
+            tmp = numerix.zeros((self._shape[0],), 'd')
+            tmp[:] = vector
+            self.put(tmp, ids, ids)
+        else:
+            ids = numerix.arange(len(vector))
+            self.put(vector, ids, ids)
+
+    def take(self, id1, id2):
+        # FIXME: This is a terrible implementation
```
Comment (tkphd) [inline fipy/matrices/petscMatrix.py]: LOL

File: fipy/matrices/petscMatrix.py [hunk 4]
```diff
@@ -0,0 +1,805 @@
...
+        v00  v01 (v02)        processor 0
+        v10  v11 (v12)
+        
+            (v01) v02 v03     processor 1
+            (v11) v12 v13
+            
+        where i is the vector index and j is the global index.
+        Elements in () are ghosted
+        
+        We end up with the GhostVec
```
Comment (tkphd) [inline fipy/matrices/petscMatrix.py]: Documentation greatly appreciated. Nice work.

File: fipy/solvers/petsc/linearLUSolver.py (hunk 1 above)
Comment (guyer) [reply fipy/solvers/petsc/linearLUSolver.py]: That's an "I wrote that note [7 years ago](https://github.com/guyer/fipy/blame/ticket644-Add_support_for_PETSc_solvers/fipy/solvers/petsc/linearLUSolver.py#L38)-TODO"

File: examples/diffusion/steadyState/mesh20x20/modifiedMeshInput.py [hunk 5]
```diff
@@ -16,12 +16,13 @@
...
 >>> valueRight = 1.
 
+>>> Lx = 20
 >>> mesh = Gmsh2D('''
 ...     cellSize = 0.5;
 ...     Point(2) = {0, 0, 0, cellSize};
-...     Point(3) = {20, 0, 0, cellSize};
-...     Point(4) = {20, 20, 0, cellSize};
-...     Point(5) = {0, 20, 0, cellSize};
+...     Point(3) = {%(Lx)g, 0, 0, cellSize};
```
[the remark this answers is not in the document]
Comment (guyer) [reply examples/diffusion/steadyState/mesh20x20/modifiedMeshInput.py]: Aw, shucks

**Comment by guyer** [comment]
> I had to add `export OMPI_MCA_plm_rsh_agent=/usr/bin/ssh` to build the docs.

@wd15: What happens if you don't?!?

File: fipy/meshes/gmshMesh.py [hunk 6]
```diff
@@ -105,8 +113,12 @@ def openMSHFile(name, dimensions=None, coordDimensions=None, communicator=parall
         Dimension of mesh
     coordDimensions : int
         Dimension of shapes
-    order
-        ???
+    overlap : int
+        The number of overlapping cells for parallel
+        simulations. Generally 1 is adequate. Higher order equations or
+        discretizations require more. If `overlap` is greater than one,
+        commumication reverts to serial, as Gmsh only provides one layer
```
[the remark this answers is not in the document]
Comment (guyer) [reply fipy/meshes/gmshMesh.py]: Fixed. [pyspelling](https://facelessuser.github.io/pyspelling/) doesn't see this because of #683.

**Comment by wd15** [comment]
I'm using petsc4py 3.8.1 and it seems to be having problems.

```Doctest: fipy.terms.cellTerm.CellTerm._test ... ok
_buildMatrix (fipy.terms.abstractDiffusionTerm._AbstractDiffusionTerm)
Doctest: fipy.terms.abstractDiffusionTerm._AbstractDiffusionTerm._buildMatrix ... /home/wd15/git/fipy-other/fipy/meshes/uniformGrid2D.py:417: DeprecationWarning: Non length-one string passed in for the array ordering. Please pass in 'C', 'F', 'A', or 'K' instead
  return numerix.concatenate((Hcen.reshape((2, self.numberOfHorizontalFaces), order="FORTRAN"),
/home/wd15/git/fipy-other/fipy/meshes/uniformGrid2D.py:420: DeprecationWarning: Non length-one string passed in for the array ordering. Please pass in 'C', 'F', 'A', or 'K' instead
  order="FORTRAN")), axis=1) + self.origin
...
barrow:30379] [29] /nix/store/c2n0xp0j3nacr1l17lmrdzwp16ljsvll-python3-3.7.5/lib/libpython3.7m.so.1.0(+0x6b53b)[0x7f5c3c54c53b]
[barrow:30379] *** End of error message ***
Aborted (core dumped)
```

Is this likely because of an older version of Petsc? I'm going to try to update to a newer version of petsc4py. Petsc4py version 3.8.4 works with the current Nixpkgs version of Petsc.

**Comment by guyer** [comment]
@wd15 & @tkphd: I'd appreciate a read-through of the scaling and threading discussion in USAGE.rst

**Comment by wd15** [comment]
I couldn't find any typos in the text you've written. All the links seemed to work as well.

The only thing is the graph. 

I would make clear what the numbers mean in the legend. In the caption write something like "Scaling behaviour of different solver packages with both varying tasks and varying threads. The number of tasks (processes) varies along the x-axis and the number of threads varies via the line style (see legend)."

The bullet points are an explanation for the behavior in the graph, but you don't describe the behavior. For example, I think you should say "Trilinos with two threads is faster than with one thread until the number of tasks reaches 10. We don't fully understand the reason for this" etc. I find it hard to interpret graphs like that without a narrative.

File: documentation/USAGE.rst [hunk 7]
```diff
@@ -256,40 +257,73 @@ class meshes. Currently, the only remaining serial-only meshes are
...
+  that our "``...Grid...``" meshes parallelize by dividing the mesh into
+  slabs, which leads to more communication overhead than more compact
+  partitions.  The "``...Gmsh...``" meshes partition more efficiently, but
+  carry more overhead in other ways.  We'll be making efforts to improve
+  the partitioning of the "``...Grid...``" meshes in a future release.
+
+- :ref:`PETSc` and :ref:`Trilinos` have fairly comparable performance, but
+  lag :ref:`PySparse` by a considerable margin.  The :ref:`SciPy` solvers
+  are even worse.  Some of this discrepancy may be because the different
+  packages are not all doing the same thing.  The meaning of the solution
```
Comment (tkphd) [inline documentation/USAGE.rst]: clarify "not doing the same thing," like

> (*e.g.*, using different preconditioners, tolerances, or norms to compute residuals)

File: documentation/USAGE.rst [hunk 8]
```diff
@@ -256,40 +257,73 @@ class meshes. Currently, the only remaining serial-only meshes are
...
+
+- There may be a *modest* benefit, *when using a large number of cpus*, to
+  allow two to four :term:`OpenMP` threads per :term:`MPI` rank.  See
+  :ref:`THREADS_VS_RANKS` for caveats more information.
+
+These results are likely both problem and architecture dependent.  You
+should develop an understanding of the scaling behavior of your own codes
+before doing "production" runs.
+
+To confirm that :term:`FiPy` is properly configured to solve in parallel,
```
Comment (tkphd) [inline documentation/USAGE.rst]: Clumsy wording. Try

> The easiest way to confirm that :term:`Fipy` is properly configured to solve in parallel is to run one of the examples, *e.g.*,::

**Comment by guyer** [comment]
@wd15 & @tkphd: thanks. I've added the clarifications requested (within the constraints of the limitations of plot captions in sphinx)

**Comment by wd15** [comment]
Looks good now.

**Comment by guyer** [comment]
Superceded by #701

**Pull request closed without merging**